config/config.go

```diff
@@ -0,0 +1,148 @@
+	// check if file exists
+	_, err := os.Stat(filePath)
+
+	// create file if not exists
+	if os.IsNotExist(err) {
+		file, err := os.Create(filePath)
+		if err != nil {
+			return nil, err
+		}
+
+		viewerConfig, err := generateDefaultConfig(filePath)
+		if err != nil {
+			return nil, err
+		}
+		marshalOpts := prototext.MarshalOptions{Multiline: true, Indent: "\t"}
+		out, err := marshalOpts.Marshal(viewerConfig)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to encode aso_sxs_viewer config %s", err)
+		}
+		out = append(out, []byte(BrowserWindowExample)...)
+		if err := ioutil.WriteFile(filePath, out, 0644); err != nil {
+			return nil, fmt.Errorf("Failed to write aso_sxs_viewer config %s", err)
+		}
+
+		defer file.Close()
+		return &ViewerConfig{*viewerConfig}, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return validateConfig(filePath)
+}
+
+func generateDefaultConfig(configPath string) (*proto.ViewerConfig, error) {
+	viewerConfig := &proto.ViewerConfig{}
+
+	viewerConfig.Url = &DefaultURL
+	viewerConfig.CssSelector = &proto.CSSSelector{
+		Selector: &DefaultCSSSelector.Selector,
+		Position: &DefaultCSSSelector.Position,
+	}
+	viewerConfig.BrowserWindowCount = &DefaultBrowserCount
+	viewerConfig.UserDataDirPrefix = &DefaultUserDataDirPrefix
+
+	DefaultInputWindowPosition := proto.ViewerConfig_DEFAULT_BOTTOM
+	viewerConfig.InputWindowPosition = &DefaultInputWindowPosition
+
+	rootWindowLayout := &proto.Layout{
+		Width:  &DefaultRootWindowWidth,
+		Height: &DefaultRootWindowHeight,
+	}
+	viewerConfig.RootWindowConfig = &proto.RootWindowConfig{
+		Layout: rootWindowLayout,
+	}
+	return viewerConfig, nil
+}
+
+// validateConfig validates the given configFile.
+func validateConfig(configPath string) (*ViewerConfig, error) {
+	in, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read aso_sxs_viewer config %s", err)
+	}
+
+	viewerConfig := &proto.ViewerConfig{}
+
+	if err := prototext.Unmarshal(in, viewerConfig); err != nil {
+		return nil, fmt.Errorf("Failed to parse aso_sxs_viewer config %s", err)
+	}
+
+	if viewerConfig == nil {
+		return nil, fmt.Errorf("Found an empty config file, please delete the existing config and try again")
+	}
+
+	if viewerConfig.BrowserWindowCount == nil {
+		viewerConfig.BrowserWindowCount = &DefaultBrowserCount
+	} else if viewerConfig.GetBrowserWindowCount() <= 0 {
+		return nil, fmt.Errorf("A positive browser_window_count field is required in the config.textproto file")
+	}
+
+	// If either the CSSSelector and url are nil, we try to iterate through the overriden windows to check
+	// if they have a missing field which does not have a default value, and throw an error if such a window is found.
+	if CSSSelector, url := viewerConfig.GetCssSelector().GetSelector(), viewerConfig.GetUrl(); CSSSelector == "" || url == "" {
+		for _, windowsOverride := range viewerConfig.GetWindowOverrides() {
```

Comment:
Shouldn't you only loop till BrowserWindowCount?
This is fine too though ... whichever you prefer.